packages/itwin/measure-tools/src/api/DrawingTypeDataCache.ts

```diff
@@ -3,42 +3,104 @@
-    this._drawingTypeCache = [];
+    this._drawingTypeCache = new Map<IModelConnection, Map<Id64String, SheetMeasurementsHelper.DrawingTypeData[]>>();
+    this._viewportModelChangedListeners = new Map<Viewport, () => void>();
+
+    this.setupEvents();
+
+    // Populate initial viewports
+    for (const vp of IModelApp.viewManager)
+      this.addViewport(vp);
+  }
+
+  private setupEvents() {
+    // If an imodel closes, clear the cache for it
+    IModelConnection.onClose.addListener((imodel) => {
+      this._drawingTypeCache.delete(imodel);
+    });
+
+    // Listen for new viewports opening
+    IModelApp.viewManager.onViewOpen.addListener((vp) => {
+      this.addViewport(vp);
+    });
+
+    // Listen for viewports closing, this also is called when IModelApp shuts down
+    IModelApp.viewManager.onViewClose.addListener((vp) => {
+      this.dropViewport(vp);
+    });
   }
 
-  public async updateDrawingTypeCache(iModel: IModelConnection) {
-    this._sheetChangeListener.forEach((func) => {
-      func();
+  private addViewport(vp: Viewport) {
+    vp.onViewedModelsChanged.addListener((viewport) =>{
+      if (!viewport.view.isSheetView())
+        return;
+
+      void this.querySheetDrawingData(viewport.iModel, viewport.view.id);
     });
-    this._sheetChangeListener = [];
-    this._drawingTypeCache = [];
+  }
+
+  private dropViewport(vp: Viewport) {
+    const listener = this._viewportModelChangedListeners.get(vp);
+    if (listener) {
+      listener();
+      this._viewportModelChangedListeners.delete(vp);
+    }
+  }
+
+  public getSheetDrawingDataForViewport(vp: Viewport): SheetMeasurementsHelper.DrawingTypeData[] | [] {
+    if (!vp.view.isSheetView())
+      return [];
 
-    const sheetIds = new Set<string>();
+    const cache = this._drawingTypeCache.get(vp.iModel);
+    if (cache)
+      return cache.get(vp.view.id) ?? [];
+
+    return [];
+  }
 
-    for (const viewport of IModelApp.viewManager) {
-      if (viewport.view.isSheetView()) {
-        this._sheetChangeListener.push(viewport.onViewedModelsChanged.addListener(async () => this.updateDrawingTypeCache(iModel)));
-        sheetIds.add(viewport.view.id);
-      }
+  public async querySheetDrawingData(imodel: IModelConnection, viewedModelID: string): Promise<Readonly<SheetMeasurementsHelper.DrawingTypeData[]>> {
+    let cache = this._drawingTypeCache.get(imodel);
+    if (!cache) {
+      cache = new Map<Id64String, SheetMeasurementsHelper.DrawingTypeData[]>();
+      this._drawingTypeCache.set(imodel, cache);
     }
 
-    for (const id of sheetIds) {
-      this._drawingTypeCache = this._drawingTypeCache.concat(await SheetMeasurementsHelper.getSheetTypes(iModel, id));
+    let sheetData = cache.get(viewedModelID);
+    if (!sheetData) {
+      sheetData = await SheetMeasurementsHelper.getSheetTypes(imodel, viewedModelID);
+      cache.set(viewedModelID, sheetData);
+    }
+
+    return sheetData;
+  }
+
+}
+
+export class DrawingDataCacheSingleton {
```

Comment:
IMO as a matter of style, I wouldn't have a separate type with the word "singleton" in it. Just export `DrawingDataCache` and have it's constructor private, with a `public static get instance(): DrawingDataCache` getter that creates the instance if undefined.